### ISSUE TITLE
decode invalid motion

### ISSUE THREAD

**Issue by OrangeSodahub** [rainmaker22/SMART#10]
In data preprocess, you sample the timestep at interval=5 from t=5, so it's possible to sample a invalid step (with correspoinding zero positions and headings), but when inference (decode future 16 tokens one by one), did you predict any invalid motion?

https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/datasets/preprocess.py#L329

**Comment by rainmaker22** [comment]
During the training phase, we only train on data where both the current and next steps are valid. This loss masking is handled in the following code: https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/modules/agent_decoder.py#L341. In the inference phase, we perform inference for all agents that are valid in the set initial frame, and future steps are predicted for all. When calculating metrics like ADE, we mask out the invalid parts of the predicted trajectories.

**Comment by OrangeSodahub** [comment]
@rainmaker22 Thanks for reply. I'm still confused about the calculation of the predicted trajecories.

As you said, all future steps are predicted one by one and each of them is a **relative** motion token, so once you encountered a invalid token, how could you get the correct trajectory of the token just after that invalid step? I saw you used gt information when preprocess:

https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/datasets/preprocess.py#L367

but in inference, you always used `pos_a` and `head_a` of last step to construct `agent_pred_rel`:

https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/modules/agent_decoder.py#L442-L452

So how did you avoid that invalid steps (ADE calculation, and visualization), thanks!

**Comment by pengzhenghao** [comment]
I think in the training phase for a trajectory once a step the agent is invalid the future steps will be all masked out?

In inference phase you don't need to worry about this.

**Comment by rainmaker22** [comment]
> I think in the training phase for a trajectory once a step the agent is invalid the future steps will be all masked out?
> 
> In inference phase you don't need to worry about this.

That's correct. During inference, you can treat all future steps as valid to obtain the full trajectory. If you need to calculate metrics like ADE, you can simply mask out the invalid trajectory points when you calculate the metric.

**Comment by OrangeSodahub** [comment]
> I think in the training phase for a trajectory once a step the agent is invalid the future steps will be all masked out?
> 
> In inference phase you don't need to worry about this.

Thanks! so only the agents which have totally 18 valid steps will be trained? I'm not sure where the code line reflects the "the future steps will be all masked out".

For inference, I mean, if I input the training data (with invalid steps), to get the full absolute trajectory, I need to get the correct relative position from the begining and calculate step by step, so if we occur a middle invalid step, we also predict a "correct" position (though not exactly equal to gt position), hence we can successfully move to the next step, is that right?

**Comment by OrangeSodahub** [comment]
> During the training phase, we only train on data where both the current and next steps are valid. This loss masking is handled in the following code:
> 
> https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/modules/agent_decoder.py#L341
> 
> . In the inference phase, we perform inference for all agents that are valid in the set initial frame, and future steps are predicted for all. When calculating metrics like ADE, we mask out the invalid parts of the predicted trajectories.

I was wondering why you only consider that the next step needs to be valid, the validity of next step should affect the current step? I'm confused.

And in your code, you actually consider both last step and next step.. @rainmaker22

**Comment by ziyanGao** [comment]
> > During the training phase, we only train on data where both the current and next steps are valid. This loss masking is handled in the following code:
> > https://github.com/rainmaker22/SMART/blob/69c7a419d5004a762ffd3305ebda4e055bd7a0db/smart/modules/agent_decoder.py#L341
> > 
> > . In the inference phase, we perform inference for all agents that are valid in the set initial frame, and future steps are predicted for all. When calculating metrics like ADE, we mask out the invalid parts of the predicted trajectories.
> 
> I was wondering why you only consider that the next step needs to be valid, the validity of next step should affect the current step? I'm confused.
> 
> And in your code, you actually consider both last step and next step.. @rainmaker22

并不是the validity of next step should affect the current step，而是next token prediction任务中，如果下一帧的token不valid的话，loss计算是无意义的，因为没有gt token做为正确的分类类别。
It's not that the validity of the next step should affect the current step, but rather that in the next token prediction task, if the token for the next frame is invalid, the loss calculation is meaningless because there is no ground truth token to serve as the correct classification category.

**Comment by OrangeSodahub** [comment]
@ziyanGao I understand what you said, but see the cited code line, it takes both last and next step into consideration. Follow your thoughts, why should we additionally consider validity of last step?

**Comment by ziyanGao** [comment]
> @ziyanGao I understand what you said, but see the cited code line, it takes both last and next step into consideration. Follow your thoughts, why should we additionally consider validity of last step?

That's a good question. This is mainly because the calculation of some input features, such as motion_vector_a, depends on the positions of the previous token and the current token, so both frames need to be valid.

**Comment by OrangeSodahub** [comment]
Thanks I get it
